### PR TITLE
Ir ro

### PR DESCRIPTION
**Pull request by serpilliere**
`AssignBlock` and `IRBlock` can be modified by external sources, but rely on some computed mecanism, which lead to hard problems to debug.

This patch make `IRBlock` them mutable through api.
`AssignBlock` is immutable, but has api to simply manipulate it.

### CODE REVIEW

**Comment by serpilliere** [comment]
Don't merge.

**Comment by serpilliere** [comment]
dont merge

File: miasm2/ir/ir.py [hunk 1]
```diff
@@ -256,15 +258,35 @@ def __init__(self, label, irs):
...
 
-    def _get_dst(self):
-        """Find the IRDst affectation and update dst, dst_linenb accordingly"""
-        if self._dst is not None:
-            return self._dst
+    @property
+    def irs(self):
+        return self._assignments
+
+    def reset_dst(self):
```
Comment (commial) [inline miasm2/ir/ir.py]: I'm not sure to see the point of this API, if IRBlock are immutable...

File: miasm2/ir/ir.py [hunk 2]
```diff
@@ -256,15 +258,35 @@ def __init__(self, label, irs):
...
+        self._dst = None
+        self._dst_linenb = None
+
+    def is_dst_set(self):
+        return self._dst is not None
+
+    def add_assignblock(self, assignblock):
```
Comment (commial) [inline miasm2/ir/ir.py]: A "del" or "remove" is missing to complete the API.
Why not having just an API to create a new IRBlock with a list of assignments?

`AssignBlock` doesn't have these kind of API (add, update, del, ...)

File: miasm2/ir/ir.py [hunk 3]
```diff
@@ -278,25 +300,34 @@ def _get_dst(self):
         self._dst_linenb = final_linenb
         return final_dst
 
-    def _set_dst(self, value):
-        """Find and replace the IRDst affectation's source by @value"""
-        if self._dst_linenb is None:
-            self._get_dst()
+    @property
+    def dst(self):
+        """Find the IRDst affectation and update dst, dst_linenb accordingly"""
```
Comment (commial) [inline miasm2/ir/ir.py]: For me, the docstring is misleading here: this is more a dev comment than the description of what `.dst` is, that is to say the destination of the block as an `Expr` instance (external user does not have to care about cache, especially in the case of immutable object).

File: miasm2/ir/ir.py [hunk 4]
```diff
@@ -278,25 +300,34 @@ def _get_dst(self):
...
-            if dst.is_id("IRDst"):
-                new_assignblk[dst] = value
-                # Sanity check is already done in _get_dst
-                break
-        self._dst = value
-        instr = self.irs[self._dst_linenb].instr
-        self.irs[self._dst_linenb] = AssignBlock(new_assignblk, instr)
-    dst = property(_get_dst, _set_dst)
+    def set_dst(self, value):
+        """Generate a new IRBlock with a dst fixed to @value"""
```
Comment (commial) [inline miasm2/ir/ir.py]: May a warning (in the docstring) saying the previous value of `IRDst`, if any, will be overwritten.

File: miasm2/ir/ir.py [hunk 5]
```diff
@@ -256,15 +256,36 @@ def __init__(self, label, irs):
 
         assert isinstance(label, AsmLabel)
         self.label = label
-        self.irs = irs
+        self._assignments = tuple(irs)
```
Comment (commial) [inline miasm2/ir/ir.py]: May the argument can be renamed to `assignments` too, or `assignblks`

File: example/ida/graph_ir.py [hunk 6]
```diff
@@ -1,40 +1,55 @@
...
+# in IDA
+
+def label_init(self, name="", offset=None):
+    self.fixedblocs = False
+    if is_int(name):
+        name = "loc_%X" % (int(name) & 0xFFFFFFFFFFFFFFFF)
+    self.name = name
+    self.attrib = None
+    if offset is None:
+        self.offset = offset
```
Comment (commial) [inline example/ida/graph_ir.py]: This line is not very clear; `self.offset = None` seems simpler

File: miasm2/ir/ir.py (hunk 1 above)
Comment (serpilliere) [inline miasm2/ir/ir.py]: Correct: I forgot to remove it

File: miasm2/ir/ir.py (hunk 2 above)
Comment (serpilliere) [inline miasm2/ir/ir.py]: The only API used in Miasm was this one. After a double check, it is only used in one file. So I am for removing it. What about you?
By the way the `__init__` is exactly the API you describe...

File: miasm2/ir/ir.py (hunk 3 above)
Comment (serpilliere) [inline miasm2/ir/ir.py]: Correct

File: miasm2/ir/ir.py (hunk 4 above)
Comment (serpilliere) [inline miasm2/ir/ir.py]: Updated.

File: miasm2/ir/analysis.py [hunk 7]
```diff
@@ -29,44 +28,43 @@ class ira_x86_16(ir_x86_16, ira)
...
-                if n_w in symbols_init and symbols_init[n_w] == v:
+            for n_w in symbexec.symbols:
+                value = symbexec.symbols[n_w]
+                if n_w in symbols_init and symbols_init[n_w] == value:
                     continue
-                eqs.append(ExprAff(n_w, v))
+                eqs.append(ExprAff(n_w, value))
             print '*' * 40
             print irb
             irb.irs = [AssignBlock(eqs)]
```
Comment (commial) [inline miasm2/ir/analysis.py]: Here, the `irb.irs` is directly modified (now, it is r-only, right?). But this method seems to never be called... (and could actually have its main part replaced by `symbexec.to_assignblock()`)

File: miasm2/ir/ir.py [hunk 8]
```diff
@@ -256,15 +258,27 @@ def __init__(self, label, irs):
...
-        if self._dst is not None:
-            return self._dst
+    @property
+    def irs(self):
+        return self._assignments
+
+    def is_dst_set(self):
+        return self._dst is not None
+
+    def updt_assignblock(self, index, assignblock):
```
Comment (commial) [inline miasm2/ir/ir.py]: Same remark than before on `add`, `del`, ...
In addition, this API seems to be never used

**Comment by serpilliere** [comment]
Fixed!

**Pull request merged**